HULKs/hulk
# Coordinate Systems Rollout

## Introduced Changes

This PR takes the concepts introduced in #611 and applies them to the whole repository.
It was previously suggested to do this whenever we touch other pieces of code, however, I think due to the ripple effects of wrapping things like `robot_to_field`, this isn't a good idea and I'd rather do it all in one richtig draufhauen.
Since this resulted in a large number of changed lines, I will outline the most important changes here.

### Coordinate Systems

Most of the codebase now uses nalgebra types like Point, Vector, or Isometry only wrapped in `Framed` and `Transform` from #611.
I have defined all of the coordinate systems that are used in our codebase in `types::coordinate_systems`, except for those used in tests which are always defined locally within the test module.

According to @schmidma, his intention was to wrap all types that exist within a coordinate system in a `Framed` type, i.e. `Framed<Ground, Circle>`.
However, I think it makes more sense to wrap only the foreign (nalgebra) types and instead make our types generic over the Frame (where applicable) which then contain nalgebra types wrapped with the frame generic.

There are mainly two reasons for this, which I will demonstrate using the example of the `Circle` type, which previously looked like this:
```rust
pub struct Circle {
    pub center: Point2<f32>,
    pub radius: f32,
}
 ```

Firstly, when accessing members of the wrapped type `Framed<Circle>`, you would have the `.inner` indirection:
`my_circle.inner.center`, which is more verbose **and**  loses the frame, as the `center` member is not itself wrapped.
Secondly, you cannot call the inner type's member functions without the `.inner` indirection or reimplementing them by `impl Framed<Circle> { ... }`.
This would be fine for type we only sometimes want to use in their wrapped form, but this is mostly for types in our `geometry` crate. Since these should always exist within a specified coordinate system anyways, there would be no point to have separate implementations for the wrapped and non-wrapped types.

If we instead have a Circle containing a `Framed<Point2>` as its `center` member, we can simply write `my_circle.center` and get a wrapped nalgebra type.
The new circle type thus looks like this:
```rust
pub struct Circle<Frame> {
    pub center: Framed<Frame, Point2<f32>>,
    pub radius: f32,
}
```

However, integrating coordinate systems with our geometry types also makes them impossible to use without specifying a coordinate system.
Depending on how seriously we want to take this idea, this is either a blessing or a curse.

I also implemented a bunch of delegating functions specifically for specific wrapped nalgebra types, so instead of `my_vector.inner.normalize().framed()` (which also loses track of the coordinate system) you can simply write `my_vector.normalize()` which returns a normalized vector tagged with the same coordinate system.
These are implemented in separate modules within the `coordinate_systems` crate.

There is some room for interpretation with the coordinate systems, especially with regards to poses and other frames of reference that change over time.
Before this PR, I never realized that we actually have two different kinds of poses, one relative to the field and one relative to the current robot pose.
The former is simply a `Transform<Ground, Field>`, but the latter would technically be a `Transform<NewGround, OldGround>`.
However, I have used `Transform<Ground, Ground>` in this PR, because even though it reduces safety slightly, it makes dealing with different poses much easier.
Even if we did use different types here, it wouldn't provide perfect safety, as positions from different KickPose coordinate systems still shouldn't be added together, even if they are both `Framed<KickPose, Point2>`s.

### Renaming `robot_to_field` -> `ground_to_field`

We have had this issue in our code for quite a while, where we accidentally called the "Ground" coordinate system "Robot".
This PR renames all occurrences of `robot_to_field` to `ground_to_field` alongside renaming variables `..._in_robot` to `..._in_ground`.
I may have missed some occurrences of the latter case.
Also, note that there still is a "Robot" coordinate frame, its just not the same as what we used to call "Robot" and should only appear in parts of the code related to kinematics.

### Limb projector

The limb projector previously used the same (unmirrored) polygons for left and right limbs, which only happens to work because the current parameters happen to be symmetric along the X-Z plane, which at least for the foot isn't accurate.
This PR does not change the polygon parameters but does mirror them for the right side of the body along with using the proper coordinate systems for each.
I also added an image overlay in twix which renders the polygons projected into the image.

### approx_derive

This PR changes our implementation for deriving the traits from the `approx` crate to work for structs without any members.
That change was necessary to derive these traits on the marker types used for coordinate system.

Fixes rabsolute bugs

## ToDo / Known Issues

- Add doc comments to the coordinate systems explaining what they are and how they relate to each other.

## Ideas for Next Iterations (Not This PR)

- Most of the changes in this PR stem from adding `.framed()` to tag  nalgebra types with a coordinate system.
But since this is a comprehensive rollout, most nalgebra variables/members are already wrapped, so most occurences of `.framed()` follow directly after a `point![...]` or `vector![...]` macro.
This could be alleviated by creating our own macros which work like the nalgebra ones but produce framed types instead, however, I have yet to come up with good names for these.
- I have implemented a bunch of accessor/delegate functions, but there are still more that can be added to prevent `a.inner.b.framed()` scenarios.

## How to Test

Since most of the changes essentially just add compile time safety, the risk of breakage should be fairly minimal and compilability checking is already done by the CI.
However, I did make a mistake during the great wrappening, see [this commit](https://github.com/HULKs/hulk/commit/33c6c20d40394843c8a93fa9caf94ca6f9178a8f) of which there are possibly more.
I already tested this on a robot, but I think its a good idea to have others test this too.
